ComponentKit/HostingView/CKComponentHostingView.mm

```diff
@@ -126,49 +128,114 @@ - (void)setContext:(id<NSObject>)context
 {
   CKAssertMainThread();
   _pendingStateUpdates.insert({globalIdentifier, stateUpdate});
-  [self _setComponentNeedsUpdate];
+  [self _setNeedsUpdateWithMode:mode];
 }
 
 #pragma mark - Private
 
-- (void)_setComponentNeedsUpdate
+- (void)_setNeedsUpdateWithMode:(CKUpdateMode)mode
 {
-  if (!_componentNeedsUpdate) { // Avoid thrashing delegate
-    _componentNeedsUpdate = YES;
-    [self setNeedsLayout];
-    [_delegate componentHostingViewDidInvalidateSize:self];
+  if (_componentNeedsUpdate && _requestedUpdateMode == CKUpdateModeSynchronous) {
+    return; // Already scheduled a synchronous update; nothing more to do.
+  }
+
+  _componentNeedsUpdate = YES;
+  _requestedUpdateMode = mode;
+
+  switch (mode) {
+    case CKUpdateModeAsynchronous:
+      [self _asynchronouslyUpdateComponentIfNeeded];
+      break;
+    case CKUpdateModeSynchronous:
+      [self setNeedsLayout];
+      [_delegate componentHostingViewDidInvalidateSize:self];
+      break;
+  }
+}
+
+- (void)_asynchronouslyUpdateComponentIfNeeded
+{
+  if (_scheduledAsynchronousComponentUpdate) {
+    return;
+  }
+  _scheduledAsynchronousComponentUpdate = YES;
+
+  // Wait until the end of the run loop so that if multiple async updates are triggered we don't thrash.
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [self _scheduleAsynchronousUpdate];
+  });
+}
+
+- (void)_scheduleAsynchronousUpdate
+{
+  if (_requestedUpdateMode != CKUpdateModeAsynchronous) {
+    // A synchronous update was either scheduled or completed, so we can skip the async update.
+    _scheduledAsynchronousComponentUpdate = NO;
+    return;
   }
+
+  id<NSObject> modelCopy = _currentModel;
+  id<NSObject> contextCopy = _currentContext;
+  CKComponentScopeRoot *scopeRootCopy = _scopeRoot;
+  auto stateUpdatesToApply = std::make_shared<const CKComponentStateUpdateMap>(_pendingStateUpdates);
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+    const CKBuildComponentResult result = CKBuildComponent(scopeRootCopy, *stateUpdatesToApply, ^CKComponent *{
+      return [_componentProvider componentForModel:modelCopy context:contextCopy];
+    });
+    dispatch_async(dispatch_get_main_queue(), ^{
+      if (!_componentNeedsUpdate) {
+        // A synchronous update snuck in and took care of it for us.
+        return;
+      }
+
+      if (_currentModel == modelCopy
+          && _currentContext == contextCopy
+          && _scopeRoot == scopeRootCopy
+          && _pendingStateUpdates.size() == stateUpdatesToApply->size()) {
+        _pendingStateUpdates.clear();
+        _component = result.component;
+        _scopeRoot = result.scopeRoot;
+        _currentModel = modelCopy;
+        _currentContext = contextCopy;
+        _componentNeedsUpdate = NO;
+        _scheduledAsynchronousComponentUpdate = NO;
+        [self setNeedsLayout];
+        [_delegate componentHostingViewDidInvalidateSize:self];
+      } else {
+        [self _scheduleAsynchronousUpdate];
```

Comment:
Sure, but if you're doing a model update at that rate and we can't compute components fast enough, there's not much we can do right?